### PR TITLE
Added cross types to :(==). Added some tests for this.

### PR DESCRIPTION
**Pull request by ecboghiu**
I spoke with Jacopo and expanding :(==) to work with different types such as `SpinVariable` and `SpinTerm{}`, etc. should be a good exercise to learn more about how the code and Julia works. I think it works fine, but maybe some part is not really efficient/elegant. I added some tests but it's not exhaustive.

### CODE REVIEW

File: src/operators.jl [hunk 1]
```diff
@@ -66,3 +66,61 @@ end
 function Base.:(==)(a::SpinMonomial, b::SpinMonomial)
     return a.variables == b.variables
 end
+
+function Base.:(==)(term1::SpinTerm{T}, term2::SpinTerm{T}) where T
```
Comment (blegat) [inline src/operators.jl]: This function is not needed, it is already implemented in
https://github.com/JuliaAlgebra/MultivariatePolynomials.jl/blob/725642e2d159fc71ad34f119e932c0715973d017/src/comparison.jl#L43-L51

File: src/operators.jl [hunk 2]
```diff
@@ -66,3 +66,61 @@ end
...
+        return var == mon.variables[first_key]
+    end
+end
+
+function Base.:(==)(mon::SpinMonomial, var::SpinVariable)
+    return var == mon
+end
+
+function Base.:(==)(mon::SpinMonomial, term::SpinTerm{T}) where T
+    if term.coefficient == 1
```
Comment (blegat) [inline src/operators.jl]: Use `isone(coefficient(term))`

File: src/operators.jl [hunk 3]
```diff
@@ -66,3 +66,61 @@ end
...
+    # If there is more than one SpinVariable in SpinMonomial then they are
+    # unequal
+    if length(mon.variables) > 1
+        return false
+    else
+        # TODO Better way of finding the first key to compare the el of Dic?
+        # Either do what I did and get the first key, or upgrade var to a
+        # dictionary and then we can just use the method fo comparing
+        # .variables field
+        first_key = collect(keys(mon.variables))[1]
```
Comment (blegat) [inline src/operators.jl]: Use `first(keys(mon.variables))`

File: src/operators.jl [hunk 4]
```diff
@@ -66,3 +66,61 @@ end
...
+        mon1, mon2 = term1.monomial, term2.monomial
+        return mon1 == mon2
+    end
+end
+
+function Base.:(==)(var::SpinVariable, term::SpinTerm{T}) where T
+    # NOTE that 1*ax and ax are different types. One is SpinVariable{Int64}
+    # and the other is SpinVarible. Also note that ax*bx*bx=ax is of type
+    # SpinVariable{bool}.
+    if term.coefficient == 1
```
Comment (blegat) [inline src/operators.jl]: Use `isone(coefficient(term))`

File: src/operators.jl (hunk 3 above)
Comment (blegat) [reply src/operators.jl]: This corresponds to a function of the MP API. Do 
```julia
function MP.variable(mon::SpinMonomial)
    if isone(length(mon.variables))
        return first(values(mon.variables))
    else
        MP._errormono2var()
    end
See

File: src/operators.jl [hunk 5]
```diff
@@ -66,3 +66,61 @@ end
...
+    # unequal
+    if length(mon.variables) > 1
+        return false
+    else
+        # TODO Better way of finding the first key to compare the el of Dic?
+        # Either do what I did and get the first key, or upgrade var to a
+        # dictionary and then we can just use the method fo comparing
+        # .variables field
+        first_key = collect(keys(mon.variables))[1]
+        return var == mon.variables[first_key]
```
Comment (blegat) [inline src/operators.jl]: Then you can just call `variable(mon)` here

File: src/operators.jl [hunk 6]
```diff
@@ -66,3 +66,61 @@ end
...
+end
+
+function Base.:(==)(term::SpinTerm{T}, var::SpinVariable) where T
+    return var == term
+end
+
+function Base.:(==)(var::SpinVariable, mon::SpinMonomial)
+    # If there is more than one SpinVariable in SpinMonomial then they are
+    # unequal
+    if length(mon.variables) > 1
```
Comment (blegat) [inline src/operators.jl]: This is also `false` if the length is zero, you should use `!isone(mon.variables)` here

**Comment by blegat** [comment]
With https://github.com/JuliaAlgebra/MultivariatePolynomials.jl/pull/140, you should be able to get away with only implementing `MP.variable`.

File: src/operators.jl [hunk 7]
```diff
@@ -66,3 +66,53 @@ end
...
+        return var == term.monomial
+    else
+        return false
+    end
+end
+
+function Base.:(==)(term::SpinTerm{T}, var::SpinVariable) where T
+    return var == term
+end
+
```
Comment (blegat) [inline src/operators.jl]: Only this method is needed, the other ones added in this PR can be removed as they are equivalent to the default fallback implemented in MultivariatePolynomials.

**Comment by blegat** [comment]
I meant keeping `MP.variable`.

**Comment by blegat** [comment]
Can you rebase the PR on master? It has conflicts

**Comment by ecboghiu** [comment]
> Can you rebase the PR on master? It has conflicts

So, rebasing was quite confusing, I think I managed to do it, but there's a good chance I did some weird stuff. Let me know if it works, if not, since my changes are really small I'll create a new branch with just the new function and do a new pull request.

**Comment by blegat** [comment]
It does not seem to have worked looking at the diff.

**Comment by ecboghiu** [comment]
I will close this one and do a new clean pull request.

**Pull request closed without merging**